Support for files over 2GB on Windows

otegami opened the pull request:
## Problem

File::Stat fails with RuntimeError when handling files >= 2GB on Windows. The stat operation crashes when
attempting to get file statistics for large files, making it impossible to work with files over 2GB.

## Cause

Windows `stat()` function uses 32-bit integers for file sizes by default, causing truncation for files >= 2GB (2^31 bytes). The regular `stat()` function and struct stat are limited to 32-bit file sizes on Windows, while Unix systems typically use 64-bit by default.

ref: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/stat-functions?view=msvc-170#time-type-and-file-length-type-variations-of-_stat

## Solution

Use Windows 64-bit stat functions and structures:
- Replace `stat()` with `_stat64()` for file statistics
- Replace `fstat()` with _fstat64() for file descriptor statistics
- Replace struct stat with struct `_stat64` for data structures
- Define macros (STAT, LSTAT, FSTAT, STAT_STRUCT) to use 64-bit variants on Windows while keeping original functions on Unix

This enables proper handling of files >= 2GB on Windows platforms while maintaining compatibility with other operating systems.

otegami committed:
Support for files over 2GB on Windows

## Problem

File::Stat fails with RuntimeError when handling
files >= 2GB on Windows. The stat operation crashes when
attempting to get file statistics for large files,
making it impossible to work with files over 2GB.

## Cause

Windows `stat()` function uses 32-bit integers for file
sizes by default, causing truncation for files >= 2GB
(2^31 bytes). The regular `stat()` function and struct
stat are limited to 32-bit file sizes on Windows, while
Unix systems typically use 64-bit by default.

ref: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/stat-functions?view=msvc-170#time-type-and-file-length-type-variations-of-_stat

## Solution

Use Windows 64-bit stat functions and structures:
- Replace `stat()` with `_stat64()` for file statistics
- Replace `fstat()` with _fstat64() for file descriptor
  statistics
- Replace struct stat with struct `_stat64` for data
  structures
- Define macros (STAT, LSTAT, FSTAT, STAT_STRUCT) to
  use 64-bit variants on Windows while keeping
  original functions on Unix

This enables proper handling of files >= 2GB on
Windows platforms while maintaining compatibility with
other operating systems.

ksss reviewed (commented):
@otegami 
Thank you for the suggestion. I agree with the policy of supporting large file sizes on Windows.

I'm a bit concerned about the `STAT_STRUCT` macro in terms of overall consistency. Instead of defining `STAT_STRUCT`, how about defining a structure like `mrb_stat`, so that we can write `struct mrb_stat`, and let `mrb_stat` branch into platform-specific structures? (Although I am a little worried that the name is too similar to `mrb_state`...)

otegami committed:
Replace STAT_STRUCT macro with mrb_stat struct name

Improve code consistency by replacing the STAT_STRUCT macro with
a more conventional mrb_stat struct name approach:

- Windows: mrb_stat -> _stat64 (for 64-bit file support)
- Unix/Linux: mrb_stat -> stat (standard struct)

otegami committed:
Use target_size directly in large file test for readability

otegami commented:
@ksss 

fix: [34be575](https://github.com/ksss/mruby-file-stat/pull/36/commits/34be575c3d0bd50ae1edee6588af4c18d1a5f0ec) Thank you for reviewing! I understand your concerns and have followed your suggestion to define a new struct to absorb the differences between platforms.

Regarding the name, How about using `mrb_file_stat`? It's a bit longer, but it would be more descriptive and avoid any confusion with `mrb_state`. What do you think?

ksss commented:
LGTM.

Since I think it will be easier to refer to if we align with the implementation in **ruby/ruby**, I believe it’s fine to keep the name as `mrb_stat`.

otegami committed:
Use typedef instead of define for type aliasing

otegami commented on src/file-stat.c:
```diff
@@ -101,11 +101,20 @@
+#  define STAT(p,s) _stat64(p,s)
+#  define FSTAT(fd,s) _fstat64(fd,s)
+#  define LSTAT(p,s) _stat64(p,s)
+#  define mrb_stat _stat64
```
fix: [7a7009e](https://github.com/ksss/mruby-file-stat/pull/36/commits/7a7009ee32a4cd0f634b7098fb795789ec85e609) Thank you so much. I got it.

ksss commented:
@otegami @kou Thank you for the improvement 🚀
